Basic jitter buffer

Hajto opened the pull request:
Adds basic functionality of jitter buffer

Hajto commented on lib/rtp_jitter_buffer.ex:
```diff
@@ -0,0 +1,89 @@
+          }
+  end
+
+  # TODO recalculate bytes to buffers
```
Should I also forward demand in bytes?

mat-hek replied:
I think it is not necessary for now

mat-hek commented on lib/rtp_jitter_buffer/cache.ex:
```diff
@@ -0,0 +1,231 @@
+defmodule Membrane.Element.RTP.JitterBuffer.Cache do
```
The name 'Cache' is quite problematic because this is not a cache.

Hajto commented on lib/rtp_jitter_buffer.ex:
```diff
@@ -0,0 +1,83 @@
+                spec: pos_integer(),
+                description: """
+                Number of slots for buffers. Each time last slot is filled `JitterBuffer`
+                will start send buffer through `:output` pad.
```
That is not correct. It does not `Each time last slot is filled will first buffer send buffer` I would rather say `Each time last slot is filled will send buffer`

mat-hek replied:
'Each time last slot is filled will first buffer send buffer' ? That's not what I suggested. Anyway, at least remove the word 'start'

Hajto commented on lib/rtp_jitter_buffer/cache.ex:
```diff
@@ -0,0 +1,231 @@
+  @type get_buffer_error :: :not_present
+
+  @doc """
+  Calls `insert_buffer/4` with metadata extracted from `Membrane.Buffer` passed as second argument.
```
That is a relict of previous versions of Cache, but...

There are two "parallel" cases of `insert_buffer/2`. They both would have `extract_secret_num/1` as their first expression and I prefer to not duplicate code. I can make `insert_buffer/2` a private helper If you think that would a better solution :)

mat-hek replied:
Not sure if not replicating a single call is worth it, but ok. Anyway, a common pattern is to leave `insert_buffer/2` public and change `insert_buffer/3` to a private counterpart `do_insert_buffer/3`

Hajto commented on lib/rtp_jitter_buffer/cache.ex:
```diff
@@ -0,0 +1,231 @@
+
+      {:ok, {record, updated_cache}}
+    else
+      {:error, :not_present}
```
Due to the fact In the first implementation of the buffer, it was driven by an external clock I've put the logic of handling missing packets into the Jitter Buffer Element itself.

It is implemented here: https://github.com/membraneframework/membrane-element-rtp-jitter-buffer/blob/basic-jitter-buffer/lib/rtp_jitter_buffer.ex#L74

And will be tested here: https://github.com/membraneframework/membrane-element-rtp-jitter-buffer/blob/basic-jitter-buffer/test/rtp_jitter_buffer/rtp_jitter_buffer_test.exs#L55

Hajto commented on lib/rtp_jitter_buffer/cache.ex:
```diff
@@ -0,0 +1,231 @@
+          {:ok, t()} | {:error, insert_error()}
+  def insert_buffer(%__MODULE__{last_seq_num: nil} = cache, buffer, seq_num) do
+    updated_cache =
+      %__MODULE__{cache | last_seq_num: seq_num - 1, end_seq_num: seq_num}
```
This is desired because next value to push will be 0 :)

Hajto replied:
Holding next to push value might be better solution. It will need a lot of changes though.

mat-hek replied:
Ok so leave it at least for now

mat-hek commented on lib/rtp_jitter_buffer/cache.ex:
```diff
@@ -0,0 +1,231 @@
+      is_fresh_packet?(last_seq_num, seq_num) ->
+        record = CacheRecord.new(buffer, seq_num)
+
+        updated_cache =
```
I haven't heard this is Erlang convention, but at least in Elixir I have seen such overriding multiple times. Googling returns this thread from the Elixir forum as the first result: https://elixirforum.com/t/is-it-ok-to-override-a-variable-or-should-i-create-a-new-one/7648 which states it is nothing wrong with it. I also cannot imagine any potential problems

mat-hek commented on lib/rtp_jitter_buffer/cache.ex:
```diff
@@ -164,68 +159,47 @@ defmodule Membrane.Element.RTP.JitterBuffer.Cache do
-  defp calc_next_seq_num(seq_number)
-  defp calc_next_seq_num(@last_seq_num), do: 0
-  defp calc_next_seq_num(seq_number) when seq_number in 0..(@last_seq_num - 1), do: seq_number + 1
+  defp pour_rollover(%__MODULE__{rollover: rollover}),
```
Is there any guarantee that rollover is initialized here?

Hajto replied:
Rollover is initialised here: https://github.com/membraneframework/membrane-element-rtp-jitter-buffer/pull/1/files/679bed843c3e5184439459850837963da4cb6995..911cfb8c53c93d1993ae8b918033d1b17c1b23e6#diff-44f5a64fe5a2cd62fa7839b146c145f2R166

mat-hek commented on lib/rtp_jitter_buffer/cache.ex:
```diff
@@ -135,25 +127,28 @@ defmodule Membrane.Element.RTP.JitterBuffer.Cache do
 
-  def skip_buffer(%__MODULE__{last_seq_num: last} = cache),
-    do: {:ok, bump_last_seq_num(cache, calc_next_seq_num(last))}
+  def dump(nil), do: []
```
This clause shouldn't be public

mat-hek commented on lib/rtp_jitter_buffer/cache.ex:
```diff
@@ -135,25 +127,28 @@ defmodule Membrane.Element.RTP.JitterBuffer.Cache do
-  defp has_rolled_over?(last_seq_num, seq_num)
-       when last_seq_num > @last_seq_num - @seq_num_rollover_delta and
+  defp has_rolled_over?(prev_seq_num, seq_num)
+       when prev_seq_num > @max_seq_number - @seq_num_rollover_delta and
```
you can just return result of this guard and remove all other clauses

mat-hek commented on lib/rtp_jitter_buffer/cache.ex:
```diff
@@ -164,68 +159,47 @@ defmodule Membrane.Element.RTP.JitterBuffer.Cache do
-    %__MODULE__{cache | heap: heap, rollover: []}
-  end
+  defp extract_seq_num(%Buffer{} = buffer),
+    do: buffer ~> (%Buffer{metadata: %{rtp: %{sequence_number: seq_num}}} -> seq_num)
```
Why don't just match in the header?

mat-hek replied:
or even better do `buffer.metadata.rtp.sequence_number`

Hajto replied:
Previously line was too long.

mat-hek commented on lib/rtp_jitter_buffer/cache.ex:
```diff
@@ -0,0 +1,205 @@
+            rollover: nil
+
+  @type t :: %__MODULE__{
+          prev_seq_num: Types.sequence_number() | nil,
```
`Types` module has been removed. Update all occurrences

Hajto commented on .vscode/launch.json:
```diff
@@ -0,0 +1,29 @@
+{
```
I am using the `.gitignore` that is used by all other elements and it is explicilty unignored:

https://github.com/membraneframework/membrane-element-rtp-jitter-buffer/pull/1/files#diff-a084b794bc0759e7a6b77810e01874f2R123

mat-hek commented on lib/rtp_jitter_buffer.ex:
```diff
@@ -39,25 +40,40 @@ defmodule Membrane.Element.RTP.JitterBuffer do
+      cache
+      |> Cache.dump()
+      |> Enum.flat_map(fn %Cache.CacheRecord{buffer: buffer} ->
+        [buffer: {:output, buffer}]
```
Nope, `buffer: {:output, [buffer1, buffer2 ...]}`

Hajto commented on lib/rtp_jitter_buffer.ex:
```diff
@@ -39,25 +40,40 @@ defmodule Membrane.Element.RTP.JitterBuffer do
+    actions =
+      cache
+      |> Cache.dump()
+      |> Enum.flat_map(fn %Cache.CacheRecord{buffer: buffer} ->
```
It was returning keyword list. So it was list of keyword lists, that needed to be flatened. Thus flat map. But I agree that returning single action with a lot of buffers is better.

mat-hek commented on lib/rtp_jitter_buffer/buffer_store.ex:
```diff
@@ -0,0 +1,200 @@
+
+  When Store is uninitialized it will accept any buffer as if it arrived on time.
+
+  Every subsequent buffer must have sequence number bigger then previous one
```
'Bigger than the _previously returned_ one'

mat-hek commented on lib/rtp_jitter_buffer/buffer_store.ex:
```diff
@@ -0,0 +1,200 @@
+    end
+  end
+
+  defp is_fresh_packet?(prev_seq_num, seq_num), do: seq_num > prev_seq_num
```
I think this should always return true until we return any buffer. Imagine first packets have sequence numbers eg. 2, 3, 1. Then the first packet is discarded (because of `prev_seq_num == 1`) while there is no reason for it, as packet 2 hasn't been returned yet

Hajto replied:
Now packet is initialized once the first packet is returned.

Hajto commented:
> a.d. [#1 (comment)](https://github.com/membraneframework/membrane-element-rtp-jitter-buffer/pull/1#discussion_r248323091)
> I know, but is there guarantee that anything has been added to the rollover before?

Nope. It was a bug in fact. I fixed it and added suitable test. Thanks for notise.

mat-hek commented on lib/rtp_jitter_buffer/buffer_store.ex:
```diff
@@ -157,7 +159,7 @@ defmodule Membrane.Element.RTP.JitterBuffer.BufferStore do
 
   defp add_to_rollover(%__MODULE__{rollover: nil} = store, %Buffer{} = buffer),
-    do: add_to_rollover(%__MODULE__{store | rollover: %__MODULE__{prev_seq_num: -1}}, buffer)
+    do: add_to_rollover(%__MODULE__{store | rollover: %__MODULE__{}}, buffer)
```
If we have that `pour_rollover`, use it instead of this clause

mat-hek commented on lib/rtp_jitter_buffer/buffer_store.ex:
```diff
@@ -0,0 +1,198 @@
+  defp add_to_rollover(%__MODULE__{rollover: rollover} = store, %Buffer{} = buffer) do
+    rollover
+    |> insert_buffer(buffer)
+    ~>> ({:ok, result} -> %__MODULE__{store | rollover: result})
```
`~>` operator should be used here

Hajto replied:
Why? Insert buffer theoretically can fail. If anything I would rather handle it [here](https://github.com/membraneframework/membrane-element-rtp-jitter-buffer/pull/1/files/64afcc7b49e002429c2069686ebbe374e0282973#diff-dd3133bc236a912e407a4a6acc8002e7R138)

Hajto replied:
Insert Buffer can fail. I changed it so if it fails insertion to store it will propagate error.
